espnet2/enh/nets/tasnet.py

```diff
@@ -0,0 +1,573 @@
+    """Reconstructs a signal from a framed representation.
+
+    Adds potentially overlapping frames of a signal with shape
+    `[..., frames, frame_length]`, offsetting subsequent frames by `frame_step`.
+    The resulting tensor has shape `[..., output_size]` where
+
+        output_size = (frames - 1) * frame_step + frame_length
+
+    Args:
+        signal: A [..., frames, frame_length] Tensor.
+                All dimensions may be unknown, and rank must be at least 2.
+        frame_step: An integer denoting overlap offsets.
+                    Must be less than or equal to frame_length.
+
+    Returns:
+        A Tensor with shape [..., output_size] containing the
+         overlap-added frames of signal's inner-most two dimensions.
+        output_size = (frames - 1) * frame_step + frame_length
+
+    Based on https://github.com/tensorflow/tensorflow/blob/r1.12/
+            tensorflow/contrib/signal/python/ops/reconstruction_ops.py
+    """
+    outer_dimensions = signal.size()[:-2]
+    frames, frame_length = signal.size()[-2:]
+
+    subframe_length = math.gcd(frame_length, frame_step)  # gcd=Greatest Common Divisor
+    subframe_step = frame_step // subframe_length
+    subframes_per_frame = frame_length // subframe_length
+    output_size = frame_step * (frames - 1) + frame_length
+    output_subframes = output_size // subframe_length
+
+    # subframe_signal = signal.view(*outer_dimensions, -1, subframe_length)
+    subframe_signal = signal.view(
+        outer_dimensions[0], outer_dimensions[1], -1, subframe_length
+    )
+
+    frame = torch.arange(0, output_subframes).unfold(
+        0, subframes_per_frame, subframe_step
+    )
+    frame = signal.new_tensor(frame).long()  # signal may in GPU or CPU
+    frame = frame.contiguous().view(-1)
+
+    # result = signal.new_zeros(*outer_dimensions, output_subframes, subframe_length)
+    result = signal.new_zeros(
+        outer_dimensions[0], outer_dimensions[1], output_subframes, subframe_length
+    )
+    result.index_add_(-2, frame, subframe_signal)
+    # result = result.view(*outer_dimensions, -1)
+    result = result.view(outer_dimensions[0], outer_dimensions[1], -1)
+    return result
+
+
+def remove_pad(inputs, inputs_lengths):
+    """
+    Args:
+        inputs: torch.Tensor, [B, C, T] or [B, T], B is batch size
+        inputs_lengths: torch.Tensor, [B]
+    Returns:
+        results: a list containing B items, each item is [C, T], T varies
+    """
+    results = []
+    dim = inputs.dim()
+    if dim == 3:
+        C = inputs.size(1)
+    for input, length in zip(inputs, inputs_lengths):
+        if dim == 3:  # [B, C, T]
+            results.append(input[:, :length].view(C, -1).cpu().numpy())
+        elif dim == 2:  # [B, T]
+            results.append(input[:length].view(-1).cpu().numpy())
+    return results
+
+
+class TasNet(AbsEnhancement):
+    def __init__(
+        self,
+        fs: int = 8000,
+        N: int = 256,
```

Comment:
as I mentioned in the config file, it is better to follow the other variable naming.

Reply:
See it. I will refactor this part.